subprojects/core-api/src/main/java/org/gradle/api/artifacts/ResolutionFailure.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.artifacts;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+/**
+ * Describes a resolution failure, giving access to context about the failure, when available.
+ * @param <T> the component id type
+ */
+@Incubating
+public interface ResolutionFailure<T extends ComponentIdentifier> {
+    /**
+     * Identifies the component that couldn't be resolved, or triggerred the resolution failure.
+     * @return The component id on which the failure occurred, if applicable.
+     */
+    @Nullable
+    T getComponentId();
+
+    /**
+     * Returns the list of locations that were tried, or null if it doesn't apply.
+     * @return the list of attempted locations
+     */
+    @Nullable
+    List<String> getAttemptedLocations();
+
+    /**
+     * The underlying exception, to be typically displayed to the advanced user.
+     *
+     * @return the problem, as an exception that might be inspected for further details.
+     */
+    Throwable getProblem();
```

Comment:
Because there are quite a few places where the component id isn't known when the exception is thrown, or where exceptions are wrapped or caught (or let go). One example: https://github.com/melix/gradle/blob/f67e7b9c28ffa0d6ca58066c1c54634ffb7e95c2/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransforms.java#L94-L94 (we catch `Throwable` because _any_ kind of exception can occur, including things we don't throw ourselves).

Another reason is that the (current `Throwable`) failures are collected from `UnresolvedDependency`. So, the initial idea was to use `UnresolvedDependency` but since the exceptions come from there, it doesn't work either. In other words, if the exception carries the context, then once it's been collected, we need to assign it its `UnresolvedDependency`, which itself contains the error, creating a nasty cycle. Keeping the exceptions on one side, and attaching them to _some context_, even if it conveys some duplication, felt cleaner to me.